### PR TITLE
[GCE] Fix ILB sharing and GC 

### LINKED ISSUE #47092: [GCE] Shared backend services for ILBs leak settings to other ILBs
PR #46663 GCP internal load balancers for services with type LoadBalancer and an annotation specifying "internal".  If the load balancer has `externalTrafficPolicy=Global`, we generate a backend service name such as `k8s-{clusterid}-internal-[tcp/udp]-[clientip|noaffinity]`. This allows other ILBs to utilize the same backend service to keep stay below quota. 

However, there are more service settings like `TimeoutSec` and `IAP` which could be unique to a single or multiple load balancers. We need to support users changing these settings (which may require modifying the backend service name format) prior to 1.7 release.  Once the release occurs, it'll be immensely difficult to change names.

### PR DESCRIPTION
**Pull request by nicksardo**
Fixes #47092 

- Users must opt-in for sharing backend services (alpha feature - may be removed in future release)
- Shared backend services use a hash for determining similarity via settings (so far, only sessionaffinity) (again, this may be removed)
- Move resource cleanup to after the ILB setup.

/assign @bowei 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce_annotations.go [hunk 1]
```diff
@@ -49,3 +54,12 @@ func GetLoadBalancerAnnotationType(service *v1.Service) (LoadBalancerType, bool)
 		return v, false
 	}
 }
+
+func GetLoadBalancerAnnotationBackendShare(service *v1.Service) bool {
```
Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/gce_annotations.go]: Add comment

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go [hunk 2]
```diff
@@ -52,19 +52,26 @@ func (gce *GCECloud) ensureInternalLoadBalancer(clusterName, clusterID string, s
...
+	}
+
 	// Lock the sharedResourceLock to prevent any deletions of shared resources while assembling shared resources here
 	gce.sharedResourceLock.Lock()
 	defer gce.sharedResourceLock.Unlock()
 
 	// Ensure health check for backend service is created
-	// By default, use the node health check endpoint
-	hcName := makeHealthCheckName(loadBalancerName, clusterID, shared)
+	// The health check is shared unless
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: finish the sentence?

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go [hunk 3]
```diff
@@ -126,39 +141,25 @@ func (gce *GCECloud) ensureInternalLoadBalancer(clusterName, clusterID string, s
...
-			// Delete the backend service as well in case it's switching between shared, nonshared, tcp, udp.
-			var existingHCName string
-			if len(bs.HealthChecks) == 1 {
-				existingHCName = getNameFromLink(bs.HealthChecks[0])
-			}
-			if err = gce.teardownInternalBackendResources(existingBSName, existingHCName); err != nil {
-				glog.Warningf("clearExistingInternalLB: could not delete old resources: %v", err)
-			} else {
-				glog.V(2).Infof("clearExistingInternalLB: done deleting old resources")
+	if len(existingBackendService.HealthChecks) == 1 {
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: What happens if its greater than 1? (will never happen?)

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go [hunk 4]
```diff
@@ -175,8 +176,8 @@ func (gce *GCECloud) updateInternalLoadBalancer(clusterName, clusterID string, s
 	_, protocol := getPortsAndProtocol(svc.Spec.Ports)
 	scheme := schemeInternal
 	loadBalancerName := cloudprovider.GetLoadBalancerName(svc)
-	shared := !v1_service.RequestsOnlyLocalTraffic(svc)
-	backendServiceName := makeBackendServiceName(loadBalancerName, clusterID, shared, scheme, protocol, svc.Spec.SessionAffinity)
+	sharedBackend := serviceIsShared(svc)
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: just use `serviceIsShared(svc)` without extra var?

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_naming.go [hunk 5]
```diff
@@ -33,15 +35,16 @@ func makeInstanceGroupName(clusterID string) string {
...
-		return fmt.Sprintf("k8s-%s-%s-%s-%s", clusterID, strings.ToLower(string(scheme)), strings.ToLower(string(protocol)), affinity)
+		hash := sha1.New()
+
+		// For every non-nil option, hash its value. Currently, only service affinity is relevant.
+		hash.Write([]byte(string(svcAffinity)))
+
+		hashed := hex.EncodeToString(hash.Sum(nil))
+		hashed = hashed[:16]
+
+		// 3 + 1 + 16 + 1 + 8 + 1 + 3 + 16
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_naming.go]: maybe document this a bit more than just the numbers?

http://lostpedia.wikia.com/wiki/The_Numbers

**Review by bowei — COMMENTED** [review]
just a few nits

File: pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go (hunk 3 above)
Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: Should never happen unless user makes manual changes.

Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: should log a warning or send an event?

Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: Looked at the GCP docs. 

> The list of URLs to the HttpHealthCheck or HttpsHealthCheck resource for health checking this BackendService. Currently at most one health check can be specified, and a health check is required. 

I'll log a warning, but as of now, it's impossible.

Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go]: ok SGTM

**Review by bowei — COMMENTED** [review]
/lgtm

**Comment by bowei** [comment]
@dchen1107 needs 1.7 milestone for merging

**Comment by nicksardo** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by ixdy** [comment]
tagging v1.7 since #47092 is 1.7

**Comment by nicksardo** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by nicksardo** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**